### PR TITLE
[llvm-debuginfo-analyzer] Add support for WebAssembly binary format.

### PR DESCRIPTION
**Pull request by CarlosAlbertoEnciso**
Add support for the WebAssembly binary format and be able to generate logical views.

https://github.com/llvm/llvm-project/issues/69181

The README.txt includes information about how to build the test cases.

### CODE REVIEW

File: llvm/docs/CommandGuide/llvm-debuginfo-analyzer.rst
Comment (sbc100) [inline llvm/docs/CommandGuide/llvm-debuginfo-analyzer.rst:1960]: wasm64 is not standardized yet so it might be best to stick to wasm32 for the example for now.

File: llvm/test/tools/llvm-debuginfo-analyzer/WebAssembly/README.txt
Comment (sbc100) [inline llvm/test/tools/llvm-debuginfo-analyzer/WebAssembly/README.txt:4]: True, but llvm does support its own `.s` format which is very similar to wat.  Would that be useful here instead of wat?

**Review by sbc100 — COMMENTED** [review]
@dschuff

File: llvm/docs/CommandGuide/llvm-debuginfo-analyzer.rst
Comment (CarlosAlbertoEnciso) [reply llvm/docs/CommandGuide/llvm-debuginfo-analyzer.rst:1960]: Changed to use wasm32.

File: llvm/test/tools/llvm-debuginfo-analyzer/WebAssembly/README.txt
Comment (CarlosAlbertoEnciso) [reply llvm/test/tools/llvm-debuginfo-analyzer/WebAssembly/README.txt:4]: You are correct in relation to the `.s` format.

All the references to `.wat` are just for illustrative purposes and the included `.wat` files are not required or used by any of the tests. I am happy to exclude them.

The llvm-debuginfo-analyzer only supports `.wasm`.

**Comment by CarlosAlbertoEnciso** [comment]
@sbc100 Thanks for your feedback.
I updated the patch to use `--target=wasm32` for documentation and tests.

File: llvm/test/tools/llvm-debuginfo-analyzer/WebAssembly/README.txt
Comment (pogo59) [reply llvm/test/tools/llvm-debuginfo-analyzer/WebAssembly/README.txt:4]: A concern that I communicated to Carlos offline is that we should not be committing binaries with no clear way to reproduce them from source. Without the .wat files the .wasm files are opaque binaries. Now, if it's sufficient to provide .c/.cpp source and use Clang to regenerate the .wasm, I'm okay with that, but I don't ever work with WebAssembly so I don't know how we usually do this.
@dwblaikie tends to have opinions about this sort of thing, I'll defer to him.

Comment (sbc100) [reply llvm/test/tools/llvm-debuginfo-analyzer/WebAssembly/README.txt:4]: I'm fine with including a way to rebuild any wasm binaries, but why not use the `.s` format, which we use elsewhere in llvm?

We don't currently have any usage or reference to wat in llvm and I think it would consistent to keep it that way.

When writing tests for the wasm binary format we tent to use wither `.yaml` files that convert to wasm objects using `objywaml` or `.s` files that convert to wasm using `llvm-mc`.    Can either of these two formats for for you? 

If you are are starting from `.c` sources that you should be able to simplest way to generate the `.s` is simply to run clang with `-S`.

Comment (CarlosAlbertoEnciso) [reply llvm/test/tools/llvm-debuginfo-analyzer/WebAssembly/README.txt:4]: > A concern that I communicated to Carlos offline is that we should not be committing binaries with no clear way to reproduce them from source. Without the .wat files the .wasm files are opaque binaries. Now, if it's sufficient to provide .c/.cpp source and use Clang to regenerate the .wasm, I'm okay with that, but I don't ever work with WebAssembly so I don't know how we usually do this. @dwblaikie tends to have opinions about this sort of thing, I'll defer to him.

@pogo59 Thanks for your valuable input.

Comment (CarlosAlbertoEnciso) [reply llvm/test/tools/llvm-debuginfo-analyzer/WebAssembly/README.txt]: @sbc100 Thanks for your suggestions to use `obj2yaml` and `llvm-mc`.
As I have already the object files, I took the option to use `obj2yaml`.

Uploaded patch:
- Removed the .wasm and .wat files.
- Removed the README.txt as it is irrelevant now.
- Added .yaml files for the tests.

File: llvm/test/tools/llvm-debuginfo-analyzer/WebAssembly/Inputs/pr-44884-clang.yaml
Comment (sbc100) [inline llvm/test/tools/llvm-debuginfo-analyzer/WebAssembly/Inputs/pr-44884-clang.yaml:241]: Its really up to you, but I think the the .s format might be readable and editable here.  But it if you are always planning on using the `.cpp` as the source of truth I guess it doesn't matter so much?

I guess maybe using the `.s` format means depending on a little more of llvm?

If the .cpp files *are* the source of truth here perhaps include a readme/script for how to regenerate the yaml files?

**Comment by CarlosAlbertoEnciso** [comment]
I don't have any preference between using `.yaml` or `.s` files. But, as you mentioned, the `.s` format might be readable and editable.
Updated the patch:
- Replaced the `.yaml` with the `.s` files.
- Added the `README.txt` with instructions (commands) on how to generate the `.s` files from the `.cpp`.

**Comment by CarlosAlbertoEnciso** [comment]
Updated patch:
- Update the 'WasmOffset' only when the current section is 'CODE'. In all the other cases, its value is zero.

**Comment by CarlosAlbertoEnciso** [comment]
Updated patch:
- Add the `%t` prefix to the generated `wasm` files to avoid any race condition, as there are tests that use the same generated `wasm`, causing corrupted input files for the `llvm-debuginfo-analyzer`.

File: llvm/docs/CommandGuide/llvm-debuginfo-analyzer.rst
Comment (dschuff) [inline llvm/docs/CommandGuide/llvm-debuginfo-analyzer.rst:2133]: copypasta? (gcc doesn't actually support wasm, right? have you tried a use case like this?)

File: llvm/lib/DebugInfo/LogicalView/Readers/LVELFReader.cpp
Comment (dschuff) [inline llvm/lib/DebugInfo/LogicalView/Readers/LVELFReader.cpp:418]: Should the name of this file/class be `LVDWARFReader` instead of ELF?

File: llvm/test/tools/llvm-debuginfo-analyzer/WebAssembly/Inputs/definitions.h
Comment (dschuff) [inline llvm/test/tools/llvm-debuginfo-analyzer/WebAssembly/Inputs/definitions.h:7]: since this is input just for testing, the compiler will always be clang and never MSVC. we could probably just write these definitions directly into the test. unless the test is shared with other compilers or platforms?

File: llvm/docs/CommandGuide/llvm-debuginfo-analyzer.rst
Comment (CarlosAlbertoEnciso) [reply llvm/docs/CommandGuide/llvm-debuginfo-analyzer.rst:2133]: You are correct; GCC does not support `wasm`.

What the `llvm-debuginfo-analyzer` does, is to process the binary file created by GCC and creates its logical view (scopes, symbols, types, lines, etc).

File: llvm/lib/DebugInfo/LogicalView/Readers/LVELFReader.cpp
Comment (CarlosAlbertoEnciso) [reply llvm/lib/DebugInfo/LogicalView/Readers/LVELFReader.cpp:418]: That is a very good point.
I think `LVDWARFReader` would match the another reader `LVCodeViewReader` as they will reflect the type of debug information format that they are parsing.

@dwblaikie, @adrian-prantl, @pogo59: any thoughts on this proposal?

File: llvm/test/tools/llvm-debuginfo-analyzer/WebAssembly/Inputs/definitions.h
Comment (CarlosAlbertoEnciso) [reply llvm/test/tools/llvm-debuginfo-analyzer/WebAssembly/Inputs/definitions.h:7]: All the test cases (`.cpp`) used by the `llvm-debuginfo-analyzer` for the WebAssembly are described in detail in the tool documentation.
We used GCC and MSVC to create binaries using those `.cpp` files that are used for the DWARF and COFF testing.
I included the `definitions.h` just to keep consistency between documentation and tests (binaries and sources).

File: llvm/lib/DebugInfo/LogicalView/Readers/LVELFReader.cpp
Comment (pogo59) [reply llvm/lib/DebugInfo/LogicalView/Readers/LVELFReader.cpp:418]: I agree. Although the renaming should be done as its own separate NFC patch.

Comment (CarlosAlbertoEnciso) [reply llvm/lib/DebugInfo/LogicalView/Readers/LVELFReader.cpp:418]: I will prepare a separate NFC patch with the following changes:
Filename change: `LVElfReader.cpp[h]`  ->  `LVDWARFReader.cpp[h]`
Class name change: `LVELFReader`  ->  `LVDWARFReader`

**Comment by dschuff** [comment]
Thanks, this looks generally good, I just had one other general question: it looks like in this PR you've added a bunch of tests for generic DWARF issues (presumably these have just been copied over from ELF tests?). This makes sense as a way to bootstrap wasm support and verify that its behavior is as expected, but it probably doesn't make sense to duplicate all of the ELF/DWARF tests, since the implementation can hopefully always be shared. For wasm it probably makes more sense to focus on trying to catch ways that they might diverge.  Is that how you're thinking about it?

Thanks for doing this btw! A while back @aheejin were thinking that it would be nice to have this for wasm, but never got around to trying to make it work!

**Comment by CarlosAlbertoEnciso** [comment]
First of all, thanks for your valuable input. I do remember @aheejin asking if there were any plans to add WebAssembly support to the tool and raising couple of issues during the initial patches landing. Happy to do this work.

**Comment by CarlosAlbertoEnciso** [comment]
I can see that may be the added tests seems like a duplication of the tests for the generic `DWARF` issues.
The main reason for their inclusion (as you pointed out) is to use them as a way to test the initial WebAssembly support and any further development.

There is another point that I think is worth to consider. The specific tests included for the `DWARF/COFF` are binary objects created by `Clang`, `MSVC` and `GCC`. The tests included for the WebAssembly are `.s` generated by `Clang` and converted to binaries by `llvm-mc` (extra step). They are a good set to test the readers supported by the tool.

I would suggest for the initial bootstrap WebAssembly support to keep them, unless that there are another issues like extra testing time, etc. Once the initial support is in place, to work with the WebAssembly people and explore areas that are important to check for divergence.

**Comment by dschuff** [comment]
yeah that makes sense; to clarify I'm not asking you to remove these tests. I just don't think it's necessary to keep new ELF tests in lockstep with wasm going forward.

**Review by dschuff — APPROVED** [review]
I'm not any kind of expert in this codebase but LGTM for whatever it's worth.

**Comment by CarlosAlbertoEnciso** [comment]
Thanks very much for your review and approval.

**Comment by CarlosAlbertoEnciso** [comment]
@sbc100 Do you have any additional comments? Thanks.

File: llvm/docs/CommandGuide/llvm-debuginfo-analyzer.rst
Comment (aheejin) [inline llvm/docs/CommandGuide/llvm-debuginfo-analyzer.rst:1985]: Nit: In LLVM `clang -c` emits `.o` and after `wasm-ld` we usually create `.wasm`. Given that the wasm file looks like a single object, would it be better to use `.o` here and below?

File: llvm/docs/CommandGuide/llvm-debuginfo-analyzer.rst
Comment (aheejin) [inline llvm/docs/CommandGuide/llvm-debuginfo-analyzer.rst:2260]: The table looks the same as the previous one after the reference and the target switched. Is that correct?

File: llvm/include/llvm/DebugInfo/LogicalView/Readers/LVBinaryReader.h
Comment (aheejin) [inline llvm/include/llvm/DebugInfo/LogicalView/Readers/LVBinaryReader.h:152]:
```suggestion
  // EXAMPLE: DW_AT_* ATTRIBUTES
```

File: llvm/lib/DebugInfo/LogicalView/Readers/LVELFReader.cpp
Comment (aheejin) [reply llvm/lib/DebugInfo/LogicalView/Readers/LVELFReader.cpp:418]: Does this mean this whole file will be renamed, or only the wasm-related parts will be spun off as a new file? The comment looks like the whole file/class will be renamed.. Then does this class not support ELF anymore? The file heading says it supports ELF/MachO. (Should we update that too?)

File: llvm/include/llvm/DebugInfo/LogicalView/Readers/LVBinaryReader.h
Comment (aheejin) [inline llvm/include/llvm/DebugInfo/LogicalView/Readers/LVBinaryReader.h:165]: How about "WasmCodeSectionOffset`? Little more verbose but I feel it less confusing.

File: llvm/lib/DebugInfo/LogicalView/Readers/LVELFReader.cpp
Comment (aheejin) [inline llvm/lib/DebugInfo/LogicalView/Readers/LVELFReader.cpp:436]: Why is this? Can you elaborate?

File: llvm/test/tools/llvm-debuginfo-analyzer/WebAssembly/README.txt
Comment (aheejin) [inline llvm/test/tools/llvm-debuginfo-analyzer/WebAssembly/README.txt:19]: Here too. We usually use `.o` for object files (pre-linking); using `.wasm` doesn't mean it does not run though.

**Review by aheejin — COMMENTED** [review]
Thank you for working on this! And sorry for the late reply. I also appreciate your help on my previous questions!

I can't say I understand all the options and tests, but given the quirks of wasm offset being mostly the only thing that needs fixing, this looks nice to me.

File: llvm/include/llvm/DebugInfo/LogicalView/Readers/LVBinaryReader.h
Comment (CarlosAlbertoEnciso) [reply llvm/include/llvm/DebugInfo/LogicalView/Readers/LVBinaryReader.h:152]: Removed the extra whitespace.

File: llvm/include/llvm/DebugInfo/LogicalView/Readers/LVBinaryReader.h
Comment (CarlosAlbertoEnciso) [reply llvm/include/llvm/DebugInfo/LogicalView/Readers/LVBinaryReader.h:165]: I think it is a good choice.

File: llvm/docs/CommandGuide/llvm-debuginfo-analyzer.rst
Comment (CarlosAlbertoEnciso) [reply llvm/docs/CommandGuide/llvm-debuginfo-analyzer.rst:2260]: The results are correct.

They match the case described in: https://llvm.org/docs/CommandGuide/llvm-debuginfo-analyzer.html#comparison-mode

In the case of WebAssembly vs DWARF, the number of missing and added types is the same:
```
 (1) Missing Types:
  -[004]     4     {TypeAlias} 'INTEGER' -> 'int'

 (1) Added Types:
  +[003]     4     {TypeAlias} 'INTEGER' -> 'int'
```

File: llvm/lib/DebugInfo/LogicalView/Readers/LVELFReader.cpp
Comment (CarlosAlbertoEnciso) [reply llvm/lib/DebugInfo/LogicalView/Readers/LVELFReader.cpp:436]: When we detect the `DW_AT_low_pc`, the `CurrentLowPC` is updated to reflect the WebAssembly code section offset.
```
  case dwarf::DW_AT_low_pc:
      FoundLowPC = true;
      ...
      if (FoundLowPC) {
        ...
        // Consider the case of WebAssembly.
        CurrentLowPC += WasmOffset;
        ...
      }
      break;
```
When we detect the DW_AT_high_pc:
```
  case dwarf::DW_AT_high_pc:
        ...
        // Don't add the WebAssembly offset if we have seen a DW_AT_low_pc.
        CurrentHighPC =
            (FoundLowPC ? CurrentLowPC - WasmOffset : CurrentLowPC) + *Offset;
        ...
      // Consider the case of WebAssembly.
      CurrentHighPC += WasmOffset;
        ...
    break;
```
Basically if we have seen a `DW_AT_low_pc`, the `CurrentLowPC` is already increased by the `WasmOffset` value; so for the calculation of `CurrentHighPC` the value of `CurrentLowPC` must be the original value (no addendums).

**Comment by CarlosAlbertoEnciso** [comment]
> Does this mean this whole file will be renamed, or only the wasm-related parts will be spun off as a new file?

The plans are to prepare a separate NFC patch with the following changes:
Filename change: `LVElfReader.cpp[h]` -> `LVDWARFReader.cpp[h]`
Class name change: `LVELFReader` -> `LVDWARFReader`

The names will reflect the Debug Information format being processed (DWARF) and not the binary format (ELF, MachO and Wasm).

> The comment looks like the whole file/class will be renamed.. Then does this class not support ELF anymore?

It will support ELF.

> The file heading says it supports ELF/MachO. (Should we update that too?)

For the time being, we can change the file heading to say something like:

```
// This implements the LVELFReader class.
// It supports ELF, Mach-O and Wasm binary formats.

```

File: llvm/docs/CommandGuide/llvm-debuginfo-analyzer.rst
Comment (CarlosAlbertoEnciso) [reply llvm/docs/CommandGuide/llvm-debuginfo-analyzer.rst:1985]: Good point. Changing `.wasm` to `.o`.

File: llvm/test/tools/llvm-debuginfo-analyzer/WebAssembly/README.txt
Comment (CarlosAlbertoEnciso) [reply llvm/test/tools/llvm-debuginfo-analyzer/WebAssembly/README.txt:19]: Good point. Changing `.wasm` to `.o`.

**Comment by CarlosAlbertoEnciso** [comment]
@aheejin Thanks for your valuable comments. Happy to help with the WebAssembly support.

File: llvm/lib/DebugInfo/LogicalView/Readers/LVELFReader.cpp
Comment (aheejin) [reply llvm/lib/DebugInfo/LogicalView/Readers/LVELFReader.cpp]: Please correct me where I'm mistaken.

Whether `CurrentLowPC` already contains the `WasmOffset` or not, `CurrentHighPC` should be higher than `CurrentLowPC`, right? The comment says
```cpp
// High PC is an offset from LowPC.
```
So if the high PC is 50 and the low PC is 40, the offset should be 10. (without `WasmOffset`)

Now add the `WasmOffset`. Suppose `WasmOffset` is 100, and `CurrentLowPC` is 140. And if we want to calculate `CurrentHighPC`, I think it still should be just
```cpp
CurrentHighPC = CurrentLowPC + *Offset;
```
So that `CurrentHighPC` is `CurrentLowPC ` (140) + `Offset` (10) = 150.

But this equation subtracts `WasmOffset` from it, so that
```cpp
CurrentHighPC = (CurrentLowPC - WasmOffset) + *Offset;
```
Here `CurrentHighPC` will be 140 - 100 + 10 = 50, whereas `CurrentLowPC` is 140, which is, `CurrentLowPC` contains `WasmOffset`, while `CurrentHighPC` does not.

I think this is strange. But given that your tool works, I must be mistaken somewhere. Can you point out where?

Comment (CarlosAlbertoEnciso) [reply llvm/lib/DebugInfo/LogicalView/Readers/LVELFReader.cpp]: Your analysis is correct until the line where `CurrentHighPC` is calculated based on `CurrentLowPC`.
Referring again to the code for `DW_AT_high_pc`:
```
  case dwarf::DW_AT_high_pc:
        ...
      if (std::optional<uint64_t> Offset = FormValue.getAsUnsignedConstant())
        // Don't add the WebAssembly offset if we have seen a DW_AT_low_pc.
        CurrentHighPC =
            (FoundLowPC ? CurrentLowPC - WasmOffset : CurrentLowPC) + *Offset;
```
At this point:
`CurrentHighPC` is 140 - 100 + 10 = **50**, whereas `CurrentLowPC` is **140**
```
        ...
      // Consider the case of WebAssembly.
      CurrentHighPC += WasmOffset;
```
At this point:
`CurrentHighPC` will be 50 + 100 = **150**, whereas `CurrentLowPC` is **140** and their offset is **10** (with WasmOffset)
```
        ...
    break;
```
I hope this clarify as how the values are calculated.

Comment (aheejin) [reply llvm/lib/DebugInfo/LogicalView/Readers/LVELFReader.cpp]: Ah I see. Then, if we are going to add `WasmOffset` three lines later, why do we subtract it here?

**Review by aheejin — APPROVED** [review]
I can't say I understand all options and test outputs, but the part I understand, mostly the wasm offset code part, looks good to me. (I asked one more question on the existing thread but that question seems minor, and I don't want this to be blocked on that)

Thanks a lot for working on this!

**Comment by CarlosAlbertoEnciso** [comment]
@aheejin Thanks very much for all your comments.
As discussed with @dschuff, now that the WebAssembly support is available, it is a great opportunity to find areas that require specific tests. Very happy to keep helping.

File: llvm/lib/DebugInfo/LogicalView/Readers/LVELFReader.cpp
Comment (CarlosAlbertoEnciso) [reply llvm/lib/DebugInfo/LogicalView/Readers/LVELFReader.cpp]: The `CurrentLowPC (DW_AT_low_pc)` always is encoded as an address.
The `CurrentHighPC (DW_AT_high_pc)` can be encoded as an address or as a constant.

When it is a constant, its value is an offset from the `CurrentLowPC`.
That is the only case where we use the formula:
`CurrentHighPC = (FoundLowPC ? CurrentLowPC - WasmOffset : CurrentLowPC) + *Offset
`
and we need to check if the `CurrentLowPC` has been already updated to include the `WasmOffset`.

If the `CurrentHighPC` is encoded as an address, we just add the `WasmOffset`.

**Pull request merged**
